index.php

```diff
@@ -250,138 +250,215 @@
+$buttonurl = new moodle_url("index.php", array(
+    "sort" => $sort,
+    "dir" => $dir,
+    "perpage" => $perpage,
+    "page" => $page,
+    "export" => 1
+    )
+);
 $buttonstring = get_string('exportbutton', 'report_coursecompletion');
 echo $OUTPUT->single_button($buttonurl, $buttonstring);
 echo $OUTPUT->footer();
 
-function add_condition_connectors(&$data, &$where, &$params, $force_and = false) {
-    if(!empty($params)) {
-        if($force_and || (isset($data->operator) && $data->operator == 0)) {
+/**
+ * Appends AND / OR operators to data fields that are assigned to $where param.
+ * @param object $data Form data stored in session.
+ * @param string $where Where clause.
+ * @param array $params Typically stores user ID and cohort ID.
+ * @param string $forceand A nonempty value would convert an OR operator
+ * to AND.
+ */
+function add_condition_connectors(
+    &$data, &$where, &$params, $forceand) {
+    $forceand = set_false_if_empty($forceand);
+
+    if (!empty($params)) {
+        if ($forceand || (isset($data->operator) && $data->operator == 0)) {
             $where .= " AND ";
         } else {
             $where .= " OR ";
         }
     }
 }
 
+/**
+ * Generates a simple html link after constructing a new instance of moodle_url.
+ * @param string $url The URL.
+ * @param string $params These params override current params in moodle_url().
+ * or add new ones.
+ * @param string $string The contents that go between opening and closing tags.
+ * @return string HTML fragment.
+ */
 function add_link($url, $params, $string) {
     return html_writer::link(new moodle_url($url, $params), $string);
 }
 
-function process_data_field(&$data, &$where, &$params, $db_field, $field_name, $comparison, $start_group = false, $force_and = false, $end_group = false) {
-    if(isset($data->{"$field_name"}) && $data->{"$field_name"}) {
-        add_condition_connectors($data, $where, $params, $force_and);
-        $params[$field_name] = $data->{"$field_name"}.($comparison == "LIKE" ? "%" : "");
+/**
+ * Builds a SQL query for case insensitive searching from form data.
+ * @param object $data Form data stored in session.
+ * @param string $where Where clause.
+ * @param array $params Search parameter. Typically stores user ID and cohort
+ * ID in other uses.
+ * @param string $dbfield Value of a database field.
+ * @param string $fieldname Name of a database field.
+ * @param string $comparison Comparison operator or a LIKE operator.
+ * @param string|bool $startgroup Could be used to append ")" to a WHERE
+ * clause.
+ * @param string $forceand A nonempty value would convert an OR operator
+ * to AND.
+ * @param string|bool $endgroup Could be used to append ")" to a WHERE clause.
+ */
+function process_data_field(&$data, &$where, &$params, $dbfield, $fieldname,
+    $comparison, $startgroup, $forceand, $endgroup = null) {
+
+    $startgroup = set_false_if_empty($startgroup);
```

Comment:
writing a new function "set_false_if_empty" is a bit strange... just use something like

$startgroup = empty($startgroup) ? false : $startgroup;